duckduckgo/macos-browser
# configuration downloading

Task/Issue URL: https://app.asana.com/0/392891325557410/1199335934832429
Tech Design URL: 
* https://app.asana.com/0/392891325557410/1199693008077758
* https://app.asana.com/0/392891325557410/1199693008077764
* https://app.asana.com/0/392891325557410/1199693008077763
* https://app.asana.com/0/392891325557410/1199894919011669

CC:

**Description**:

Automatically downloads configuration files and updates dependencies.

Note, in debug mode it will do a remote check every 2 two minutes, in production this is 12 hours.

Application Support location: ~/Library/Containers/com.duckduckgo.macos.browser/Data/Library/Application\ Support/

**Steps to test this PR**:
1. Startup Charles proxy and ensure that you can see requests to DuckDuckGo via HTTPs
1. Run the app, a Configuration folder will be created under the Application Support location
1. *Quickly* load a couple of pages and connect to them with the web inspector.  The source of the content blocker and content blocker rules JavaScript will not include surrogates or excluded domains. Also make a note of the tracker radar content if possible
1. After a minute the update process will begin. You can use the app as normal during this time.  Check Charles for the requests, there should only be a `If-None-Match` header for the TrackerRadarData.  Responses should all be 200 and contain data.
1. Reload the pages and check the sources with the inspector again - they should now contain surrogates, excluded domains and contain an updated tracker radar list.  Also check that blocking is occurring by checking the JS console on a site with trackers.
1. Wait another 2-3 minutes, nothing should happen other than the "refresh check" debug log entry.  Check Charles again, the requests should all contain `If-None-Match` headers.  Responses should all be 302. 
1. Delete any file from the Configuration folder and wait another 2-3 minutes
1. The process should run and update the missing the file

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
